packages/expo/cli/start/server/middleware/ManifestMiddleware.ts

```diff
@@ -0,0 +1,245 @@
+  constructor(
+    protected projectRoot: string,
+    protected options: {
+      /** Should start the dev servers in development mode (minify). */
+      mode?: 'development' | 'production';
+      /** Should instruct the bundler to create minified bundles. */
+      minify?: boolean;
+      constructUrl: UrlCreator['constructUrl'];
+      isNativeWebpack?: boolean;
+    }
+  ) {
+    super(
+      projectRoot,
+      /**
+       * Only support `/`, `/manifest`, `/index.exp` for the manifest middleware.
+       */
+      ['/', '/manifest', '/index.exp']
+    );
+  }
+
+  protected getDefaultResponseHeaders(): Map<string, any> {
+    return new Map<string, any>();
+  }
+
+  /** Exposed for testing. */
+  public async _resolveProjectSettingsAsync({
+    platform,
+    hostname,
+  }: Pick<ParsedHeaders, 'hostname' | 'platform'>): Promise<ResponseProjectSettings> {
+    // Read the config
+    const projectConfig = getConfig(this.projectRoot);
+
+    // Read from headers
+    const mainModuleName = this.resolveMainModuleName(projectConfig, platform);
+
+    // Create the manifest and set fields within it
+    const expoGoConfig = this.getExpoGoConfig({
+      mainModuleName,
+      hostname,
+    });
+
+    const hostUri = this.options.constructUrl({ scheme: '', hostname });
+
+    const bundleUrl = this._getBundleUrl({
+      platform,
+      mainModuleName,
+      hostname,
+    });
+
+    // Resolve all assets and set them on the manifest as URLs
+    await this.mutateManifestWithAssetsAsync(projectConfig.exp, bundleUrl);
+
+    return {
+      expoGoConfig,
+      hostUri,
+      bundleUrl,
+      exp: projectConfig.exp,
+    };
+  }
+
+  /** Get the main entry module ID (file) relative to the project root. */
+  private resolveMainModuleName(projectConfig: ProjectConfig, platform: string): string {
+    let entryPoint = resolveEntryPoint(this.projectRoot, platform, projectConfig);
+    // NOTE(Bacon): Webpack is currently hardcoded to index.bundle on native
+    // in the future (TODO) we should move this logic into a Webpack plugin and use
+    // a generated file name like we do on web.
+    // const server = getDefaultDevServer();
+    // // TODO: Move this into BundlerDevServer and read this info from self.
+    // const isNativeWebpack = server instanceof WebpackBundlerDevServer && server.isTargetingNative();
+    if (this.options.isNativeWebpack) {
+      entryPoint = 'index.js';
+    }
+
+    return stripExtension(entryPoint, 'js');
+  }
+
+  /** Parse request headers into options. */
+  public abstract getParsedHeaders(req: ServerRequest): ParsedHeaders;
+
+  /** Store device IDs that were sent in the request headers. */
+  private async saveDevicesAsync(req: ServerRequest) {
+    const deviceIds = req.headers?.['expo-dev-client-id'];
+    if (deviceIds) {
+      await ProjectDevices.saveDevicesAsync(this.projectRoot, deviceIds).catch((e) =>
+        Log.error(e.stack)
```

Comment:
I think the error message/summary is always great to present to users, seems like that's being swallowed here right? Maybe it's better to add something [like this](https://github.com/expo/expo/pull/16559/files#diff-fb386925332777171063110d0764fb5db28ed815d9a3bd8dc9b3a55b3007b744R46) here too:
```js
Log.error(chalk.red(e.toString()) + (EXPO_DEBUG ? '\n' + chalk.gray(e.stack) : ''));
```

Or even something like `Log.exception` which does exactly that?